src/proj.rs

```diff
@@ -0,0 +1,791 @@
+    #[inline(always)]
+    fn warp_t(&self, x: f32, y: f32) -> (f32, f32) {
+        let t = &self.transform;
+        let tx = t[2];
+        let ty = t[5];
+
+        (x + tx,
+         y + ty)
+    }
+}
+
+/// Combines two projective transformations
+/// (matrix multiplication)
+impl Mul<Proj> for Proj {
+    type Output = Proj;
+
+    fn mul(self, rhs: Proj) -> Proj {
+        use TransformationClass as TC;
+        let t = mul3x3(self.transform, rhs.transform);
+        let i = mul3x3(rhs.inverse, self.inverse);
+
+        let class = match (self.class, rhs.class) {
+            (TC::Translation, TC::Translation) => TC::Translation,
+            (TC::Translation, TC::Affine) => TC::Affine,
+            (TC::Affine, TC::Translation) => TC::Affine,
+            (TC::Affine, TC::Affine)      => TC::Affine,
+            (_, _) => TC::Projection
+        };
+
+        Proj { transform: t, inverse: i, class: class }
+    }
+}
+
+/// Combines two projective transformations
+/// (matrix multiplication)
+impl<'a, 'b> Mul<&'b Proj> for &'a Proj {
+    type Output = Proj;
+
+    fn mul(self, rhs: &Proj) -> Proj {
+        use TransformationClass as TC;
+        let t = mul3x3(self.transform, rhs.transform);
+        let i = mul3x3(rhs.inverse, self.inverse);
+        let class = match (self.class, rhs.class) {
+            (TC::Translation, TC::Translation) => TC::Translation,
+            (TC::Translation, TC::Affine) => TC::Affine,
+            (TC::Affine, TC::Translation) => TC::Affine,
+            (_, _) => TC::Projection
+        };
+
+        Proj { transform: t, inverse: i, class: class }
+    }
+}
+
+impl Mul<(f32, f32)> for Proj {
+    type Output = (f32, f32);
+
+    fn mul(self, rhs: (f32, f32)) -> (f32, f32) {
+        let (x,y) = rhs;
+        self.warp_p(x, y)
+    }
+}
+
+impl<'a, 'b> Mul<&'b (f32, f32)> for &'a Proj {
+    type Output = (f32, f32);
+
+    fn mul(self, rhs: &(f32, f32)) -> (f32, f32) {
+        let (x,y) = *rhs;
+        self.warp_p(x, y)
+    }
+}
+
+/// Performs projective transformation to an image. Allocates an output image with same
+/// dimensions as the input image. Output pixels outside the input image are set to `default`.
+/// Proj `homography` defines a mapping from coordinates in the `output` image to coordinates of
+/// the `image`.
+pub fn warp_new<P>(homography: &Proj,
+                   image: &Image<P>,
+                   interpolation: Interpolation,
+                   default: P
+                  ) -> Image<P> 
+where
+P: Pixel + Send + Sync + 'static,
+<P as Pixel>::Subpixel: Send + Sync,
+<P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32>,
+{
+    let (width, height) = image.dimensions();
+    let mut out = ImageBuffer::new(width, height);
+    warp(homography, image, interpolation, default, &mut out);
+
+    out
+}
+
+/// Performs projective transformation `homography` mapping pixels from
+/// `image` into `&mut output`.
+/// Proj `homograpy` defines coordinate mapping from `out` to `image`.
```

Comment:
This seems like the wrong way round...

Reply:
@theotherphil, this is common in many image processing libraries (check out opencv's docs for example: https://docs.opencv.org/2.4/modules/imgproc/doc/geometric_transformations.html?highlight=warp).

Since getting the inverse transformation from Proj is super cheap, it could be either way. But if we want to keep functions warp_with that allow passing a custom function, than that could be rather confusing to have two different conventions in the same crate. 

What do you think?

Reply:
OpenCV isn't really the cleanest reference to be honest, I always found it super weird and piecing together transformations without the reference close to impossible. Though it is prior art...

Reply:
Maybe this can be resolve with naming:

* `warp`: takes only `Proj` but 'the right way around'. Parameter order: `(image, Proj, Interpolation, out)`
* `mapped_from`: the arbitrary mapping and thus reverse projection. other parameter order
   `(out, projection, Interpolation, image)`

Just an idea

Reply:
Not bad. I was playing with the idea, but reverted to the `accepted` order after I added warp_with.